WordPress/gutenberg
# Fix query loop bugs by correctly relying on the main query and removing problematic workaround

This PR will fix https://core.trac.wordpress.org/ticket/58027 and https://core.trac.wordpress.org/ticket/59225 once merged into WP core. The bug can be replicated with Gutenberg as well though.

## What?
Currently, block themes do not handle "the loop" correctly. Since their existence, WordPress themes have been expected to have the main `WP_Query` object go through the loop of posts when rendering posts, which is commonly done with code such as the following:
```php
while ( have_posts() ) {
	the_post();
	// Render post etc.
}
```

The above logic ensures the WordPress query loop is correctly handled, e.g. calling `in_the_loop()` will return `true` as expected.

In block themes, this doesn't happen correctly though, the loop is not started when rendering posts which breaks the long-standing expectation that is relied on in both WP core and the plugin ecosystem. This is because the `core/post-template` block, when configured to use the global query (via the `inherit` attribute), _clones_ the global query object rather than using it directly. This leads to the main query loop never being started since it is only started on the cloned instance.

Before that problem was introduced even, based on an older version of the query block which at the time also did not trigger the loop correctly, the lack of proper loop handling was "addressed" with a workaround in the `core/post-content` (and later `core/post-featured-image`) block which effectively forces the main query loop to start if it was not started before. That workaround is not a proper fix and itself causes bugs, such as the one outlined in https://core.trac.wordpress.org/ticket/58027.

Looking at the history led to the following findings:
* #43198 (via #43221) introduced the latest underlying query loop bug, in an attempt to fix another (more edge-case) bug originally reported in https://github.com/Automattic/wp-calypso/issues/66270.
* #30405 introduced the problematic workaround in the `core/post-content` block which is a hack to force the loop to start if it hasn't started yet.
* #45534 then added the same hack to the `core/post-featured-image` block.

## Why?
WordPress has always expected the query loop to be used when rendering posts, even when a template is only expected to render a single post (such as the `single` or `page` templates). Classic themes therefore start the loop even on those templates, but it looks like this was missed in block themes, which has led to the aforementioned problems.

Workarounds like enforcing the loop in certain blocks is not a proper solution for that and as mentioned above has led to its own set of problems.

## How?
This PR ensures that the loop is properly handled in block themes, in a way that fixes the bug in https://core.trac.wordpress.org/ticket/58027 while also maintaining a fix for https://github.com/Automattic/wp-calypso/issues/66270#issuecomment-1209752872.

* In block templates where the global main query is used, the `core/post-template` block is updated to _actually_ use the main query directly which will fix the loop to work correctly with any `in_the_loop()` checks and address https://core.trac.wordpress.org/ticket/59225.
    * Also in the `core/post-template` block, if there _is_ already a loop (which would only happen when using the block in a classic theme, or an even more edge-case scenario where the block is used nested within another `core/post-template` block), the cloning of the main query is maintained, however with the addition of a `WP_Query::rewind_posts()` call on the cloned instance, to ensure that, even if the loop on the main query was already started, the cloned instance will render _all_ relevant posts. This is effectively a more appropriate fix to https://github.com/Automattic/wp-calypso/issues/66270 / #43198.
* The workaround to enforce the loop in `core/post-content` and `core/post-featured-image` is removed as it's no longer necessary following the above fix, which effectively will fix the bug reported in https://core.trac.wordpress.org/ticket/58027.

## Testing Instructions
Before testing, add the following code snippet to your WP installation (could be added anywhere, in a plugin, or hacked into any core or theme functions file):
```php
add_filter(
	'render_block_core/null',
	function( $block_content ) {
		return $block_content . '<p>in the loop: ' . ( in_the_loop() ? 'true' : 'false' ) . '</p>';
	}
);
add_action( 'wp_body_open', function() {
	if ( wp_is_block_theme() ) { // Only use this code when using a classic theme.
		return;
	}
	$block_content = '';
	$block = parse_blocks($block_content)[0];
	echo render_block($block);
} );
```

1. Create 2+ dummy posts, each with a title and regular text content (e.g. `core/paragraph` blocks).
2. Use Twenty Twenty-Three (block theme).
3. Open the home page. You should see under each post "in the loop: true". This works only because of the workaround in the `core/post-content` and `core/post-featured-image` blocks.
4. Open the Site Editor and edit the home page ("Blog Home") template: Remove the `core/post-content` and `core/post-featured-image` blocks from within the post template (which should lead to the featured image placeholders and content disappearing from all posts displayed).
5. Open the home page again. Now you should see under each post "in the loop: false", confirming that the loop is no longer entered and only worked due to the workaround.
6. Switch to Twenty Twenty-One (classic theme).
7. Open any of the individual posts in the frontend (single template). You should see no post content being displayed. This is because the `core/featured-image` block is technically displayed (if any of the posts has a featured image, you should see that on top of the page), forcing the loop to start before it should. So then when the actual theme attempts to start the loop, it has already reached the end and therefore displays nothing.

Now, apply this PR to Gutenberg (or the mirror PR https://github.com/WordPress/wordpress-develop/pull/5103 to WP core), then test the above again. You don't need to create more posts of course, so you can start at step 2. You should see in steps 5 and 7 that the bugs have been fixed:
* In step 5, you should now see "in the loop: true" under each post.
* In step 7, you should now see the actual post with its title, content etc.

Last but not least, test the following to ensure the bug fixed in https://github.com/Automattic/wp-calypso/issues/66270 / #43198 remains fixed (i.e. no PHP logic infinite loop):
1. Use Twenty Twenty-One (classic theme).
2. Create a new post and add a `core/query` block to it (use any of the patterns, such as "Grid"). Make sure the "Inherit query from template" toggle is enabled in the sidebar.
3. Publish the post and view it in the frontend. You should see the post appear within the post, and it should have "in the loop: true" below.
